A67-xds-dynamic-parameters.md

```diff
@@ -0,0 +1,266 @@
+(pseudo code):
+
+```text
+// Example authority_resource_names:
+[
+  // Resources for authority "foo.example.com"
+  { resource_names: ["bar", "baz"], dynamic_parameters: {"env": "prod"} },
+
+  // Resources for authority "zzz.example.com"
+  { resource_names: ["yyy"], dynamic_parameters: {"env": "prod", "version": "v2"} },
+]
+
+function ConstructADSRequest(authority_resource_names)
+  ...
+  foreach resource_names, dynamic_parameters in authority_resource_names
+    if dynamic_parameters is empty
+      request.resource_names = resource_names
+    else
+      foreach name in resource_names
+        request.resource_locators.add(name, dynamic_parameters)
+  ...
+```
+
+### Temporary environment variable protection
+
+During development, the dynamic parameter functionality will be guarded by the
+GRPC_EXPERIMENTAL_XDS_DYNAMIC_PARAMETERS env var. Specifically, this env var
+will control two things:
+
+-   Whether we read the new `dynamic_parameters` fields in the bootstrap
+    config (and therefore whether we populate the `resource_locators` field in
+    ADS requests). This guards against cases where the bootstrap file is
+    generated by a separate tool with a release process independent of that of
+    the application (e.g., the Traffic Director bootstrap generator), such
+    that a newer bootstrap config may be used with an older version of gRPC.
+-   Whether we look at the new `resource_names` field in ADS responses. This
+    guards against xDS servers that may incorrectly start sending new fields
+    to clients that are not yet ready to understand them.
+
+The env var protection will be removed once this feature passes interop tests.
+
+## Rationale
+
+The following sub-sections describe tradeoffs that were made in this design.
+
+### No changas to dynamic parameters after startup
+
+For our initial support of dynamic parameters, we can let the parameter
+key-value pairs be only read at startup. A client restart will be required when
+changing the parameters.
+
+### No metadata reuse
+
+An idea in TP2 is to reuse the node metadata field as dynamic parameters.
+
+This idea is especially useful as a possible migration path for Envoy, where
+servers use node metadata to determine which resources to send for wildcard
+LDS and CDS subscrioptions. In contrast, node metadata is not as important in
+the gRPC xDS ecosystem, so this migration path is not considered important.
+
+Additionally, converting node metadata to dynamic parameters would be somewhat
+complex to implement, since node metadata can have structured values, whereas
+dynamic parameter values are always just strings.
+
+Additionally, an explicit knob in the bootstrap config that determines whether
+to use dynamic parameters is desirable for control. Using the presence of the
+`dynamic_parameters` field gives us that.
+
+### No merging of top-level and per-authority dynamic parmeters
+
+The `dynamic_parameters` at root level only apply to the non-xdstp case. There
```

Comment:
That is not clear in the proposal text. It just says "To support non-xdstp use cases, we can add a map at the root level too". That doesn't require that it is used exclusively in non-xdstp.

I do wonder if there's something we can do here to avoid confusion where someone thinks they are configuring the dynamic parameters for xdstp, but actually aren't. However, I see the value in using the same dynamic_parameters name, just like we are doing with xds_servers. Although for xds_servers, if you don't specify it in the authorities section the top-level xds_servers is used. Maybe we should do that; it becomes the default, but there's still no merging. That's probably the least surprising. Although it does cause problems for disabling the dynamic parameters within authorities, because empty list is semantically the same as not present.